pkg/aws-s3/csv.go

```diff
@@ -2,104 +2,314 @@
+			return fmt.Errorf("failed to read LF after CR: %w", readErr)
+		}
+
+		*bytesRead++
+
+		lineBuffer.WriteByte(nextByte)
+	}
+
+	return nil
+}
+
+func readCSVLine(reader *bufio.Reader) (
+	lineBytes []byte, bytesRead int64, err error) {
+	var lineBuffer bytes.Buffer
+
+	inQuotes := false
+
+	for bytesRead = 0; ; {
+		if bytesRead >= MaxCSVRowSizeBytes {
+			return nil, 0, fmt.Errorf("size limit of %d MB exceeded", MaxCSVRowSizeBytes/(1024*1024))
+		}
+
+		b, readErr := reader.ReadByte()
+		if readErr != nil {
+			if readErr == io.EOF {
+				if lineBuffer.Len() == 0 && bytesRead == 0 {
+					return nil, 0, ErrEmptyOrMissing
+				}
+
+				break
+			}
+
+			return nil, 0, fmt.Errorf("failed to read byte: %w", readErr)
+		}
+
+		bytesRead++
+
+		lineBuffer.WriteByte(b)
+
+		if b == '"' {
+			if err := handleQuoteChar(reader, &lineBuffer, &bytesRead, &inQuotes); err != nil {
+				return nil, 0, err
+			}
+		} else if (b == '\n' || b == '\r') && !inQuotes {
+			if b == '\r' {
+				if err := handleLineEnding(reader, &lineBuffer, &bytesRead); err != nil {
+					return nil, 0, err
+				}
+			}
+
+			break
+		}
+	}
+
+	lineBytes = lineBuffer.Bytes()
+
+	if len(lineBytes) > 0 && lineBytes[len(lineBytes)-1] == '\r' {
+		lineBytes[len(lineBytes)-1] = '\n'
+	}
+
+	return lineBytes, bytesRead, nil
+}
+
+func CSVHeaders(reader *bufio.Reader) (headers []string, bytesReadForHeader int64, err error) {
+	headerLineBytes, bytesRead, err := readCSVLine(reader)
 
-	// Read all the CSV data
-	records, err := csvData.ReadAll()
 	if err != nil {
-		return nil, false, fmt.Errorf("failed to read CSV data: %v", err)
+		return nil, bytesRead, fmt.Errorf("CSV header error: %w", err)
 	}
 
-	count := len(records)
-	if count == 0 {
-		return nil, false, fmt.Errorf("no data found in the CSV file")
+	csvReader := csv.NewReader(bytes.NewReader(headerLineBytes))
+	parsedHeaders, parseErr := csvReader.Read()
+
+	if parseErr != nil {
+		return nil, 0, fmt.Errorf("CSV file format is invalid or corrupted: %v", parseErr)
 	}
 
-	objects := make([]map[string]any, 0, pageSize)
-	if count == 1 {
-		return objects, false, nil
+	if len(parsedHeaders) == 0 {
+		return nil, 0, fmt.Errorf("CSV header error: empty or missing")
 	}
 
-	// Convert CSV data to a slice of maps
-	headers := records[0]
+	return parsedHeaders, bytesRead, nil
+}
+
+func StreamingCSVToPage(
+	streamReader *bufio.Reader,
+	headers []string,
+	pageSize int64,
+	attrConfig []*framework.AttributeConfig,
+	maxProcessingBytesTotal int64,
+) (objects []map[string]any, bytesReadFromDataStream int64, hasNext bool, err error) {
+	objects = make([]map[string]any, 0, pageSize)
 	headerToAttributeConfig := headerToAttributeConfig(headers, attrConfig)
 
-	end := min(start+pageSize, int64(count))
-	hasNext := end < int64(count)
+	var totalBytesRead int64
+
+	hasNext = true
+
+	for int64(len(objects)) < pageSize {
+		if maxProcessingBytesTotal > 0 && totalBytesRead >= maxProcessingBytesTotal {
+			break
+		}
```

Comment:
Why do we need to check if `maxProcessingBytesTotal` is greater than zero or not?

```suggestion
		if totalBytesRead >= maxProcessingBytesTotal {
			break
		}
```

Reply:
It was a precautionary check since it is a function parameter. In this case, we don't need it. Maybe we can update it later, in the future if needed.

Reply:
I think it is better to remove this. It can have some unintended consequences such as if zero is passed, then no limit is applied.

Better convert all limit types to `uint64`.

This is not a blocker and can come in a follow up PR.